File: tests/wasm/test_coppock.js
/**
 * WASM binding tests for COPPOCK indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS 
} from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
        
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('COPPOCK accuracy', () => {
    
    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.coppock;
    
    const result = wasm.coppock_js(
        close,
        expected.defaultParams.short,
        expected.defaultParams.long,
        expected.defaultParams.ma,
        expected.defaultParams.ma_type
    );
    
    assert.strictEqual(result.length, close.length);
    
    
    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected.last5Values,
        1e-8,
        "COPPOCK last 5 values mismatch"
    );
});

test('COPPOCK default parameters', () => {
    
    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.coppock;
    
    const result = wasm.coppock_js(
        close, 
        expected.defaultParams.short,
        expected.defaultParams.long,
        expected.defaultParams.ma,
        expected.defaultParams.ma_type
    );
    assert.strictEqual(result.length, close.length);
    
    
    const warmup = Math.max(expected.defaultParams.short, expected.defaultParams.long) + (expected.defaultParams.ma - 1);
    
    
    let foundNonNaN = false;
    for (let i = warmup; i < result.length; i++) {
        if (!isNaN(result[i])) {
            foundNonNaN = true;
            break;
        }
    }
    assert(foundNonNaN, "No valid values found after warmup");
});

test('COPPOCK zero period', () => {
    const inputData = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.coppock_js(inputData, 0, 14, 10, "wma");
    }, /Invalid period/);
});

test('COPPOCK period exceeds length', () => {
    const dataSmall = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.coppock_js(dataSmall, 11, 14, 10, "wma");
    }, /Invalid period|Not enough valid data/);
});

test('COPPOCK very small dataset', () => {
    const singlePoint = new Float64Array([42.0]);
    
    assert.throws(() => {
        wasm.coppock_js(singlePoint, 11, 14, 10, "wma");
    }, /Invalid period|Not enough valid data/);
});

test('COPPOCK empty input', () => {
    const empty = new Float64Array([]);
    
    assert.throws(() => {
        wasm.coppock_js(empty, 11, 14, 10, "wma");
    }, /Empty data/);
});

test('COPPOCK all NaN input', () => {
    const allNaN = new Float64Array(100);
    allNaN.fill(NaN);
    
    assert.throws(() => {
        wasm.coppock_js(allNaN, 11, 14, 10, "wma");
    }, /All values are NaN/);
});

test('COPPOCK batch single parameter', () => {
    
    const close = new Float64Array(testData.close);
    
    const batchResult = wasm.coppock_batch(close, {
        short_range: [11, 11, 0],
        long_range: [14, 14, 0],
        ma_range: [10, 10, 0],
        ma_type: "wma"
    });
    
    
    const singleResult = wasm.coppock_js(close, 11, 14, 10, "wma");
    
    assert.strictEqual(batchResult.values.length, singleResult.length);
    
    
    assertArrayClose(batchResult.values, singleResult, 1e-9, "Batch vs single mismatch");
});

test('COPPOCK batch multiple parameters', () => {
    
    const close = new Float64Array(testData.close.slice(0, 100)); 
    
    
    const batchResult = wasm.coppock_batch(close, {
        short_range: [10, 12, 2],    
        long_range: [14, 16, 2],     
        ma_range: [8, 10, 2],        
        ma_type: "wma"
    });
    
    
    assert.strictEqual(batchResult.combos.length, 8);
    assert.strictEqual(batchResult.rows, 8);
    assert.strictEqual(batchResult.cols, 100);
    assert.strictEqual(batchResult.values.length, 8 * 100);
    
    
    for (let combo = 0; combo < batchResult.combos.length; combo++) {
        const short_period = batchResult.combos[combo].short_roc_period;
        const long_period = batchResult.combos[combo].long_roc_period;
        const ma_period = batchResult.combos[combo].ma_period;
        assert(batchResult.combos[combo].ma_type === "wma", "MA type should be set");
        
        const rowStart = combo * 100;
        const rowData = batchResult.values.slice(rowStart, rowStart + 100);
        
        
        const warmup = Math.max(short_period, long_period) + ma_period - 1;
        for (let i = 0; i < Math.min(warmup, 100); i++) {
            assert(isNaN(rowData[i]), `Expected NaN at warmup index ${i}`);
        }
    }
});

test('COPPOCK zero-copy API', () => {
    
    const data = new Float64Array([
        10, 20, 30, 40, 50, 60, 70, 80, 90, 100,
        110, 120, 130, 140, 150, 160, 170, 180, 190, 200
    ]);
    
    const short_period = 5;
    const long_period = 7;
    const ma_period = 3;
    
    
    const ptr = wasm.coppock_alloc(data.length);
    assert(ptr !== 0, 'Failed to allocate memory');
    
    
    const memView = new Float64Array(
        wasm.__wasm.memory.buffer,
        ptr,
        data.length
    );
    
    
    memView.set(data);
    
    
    try {
        wasm.coppock_into(ptr, ptr, data.length, short_period, long_period, ma_period, "wma");
        
        
        const regularResult = wasm.coppock_js(data, short_period, long_period, ma_period, "wma");
        for (let i = 0; i < data.length; i++) {
            if (isNaN(regularResult[i]) && isNaN(memView[i])) {
                continue; 
            }
            assert(Math.abs(regularResult[i] - memView[i]) < 1e-10,
                   `Zero-copy mismatch at index ${i}: regular=${regularResult[i]}, zerocopy=${memView[i]}`);
        }
    } finally {
        
        wasm.coppock_free(ptr, data.length);
    }
});

test('COPPOCK zero-copy error handling', () => {
    
    assert.throws(() => {
        wasm.coppock_into(0, 0, 10, 11, 14, 10, "wma");
    }, /null pointer|Null pointer/i);
    
    
    const ptr = wasm.coppock_alloc(10);
    try {
        
        assert.throws(() => {
            wasm.coppock_into(ptr, ptr, 10, 0, 14, 10, "wma");
        }, /Invalid period/);
        
        
        assert.throws(() => {
            wasm.coppock_into(ptr, ptr, 10, 20, 14, 10, "wma");
        }, /Period exceeds data length/);
    } finally {
        wasm.coppock_free(ptr, 10);
    }
});

test('COPPOCK supported MA types', () => {
    
    const close = new Float64Array(testData.close.slice(0, 100));
    
    
    const maTypes = ['sma', 'ema', 'wma'];
    const results = {};
    
    for (const maType of maTypes) {
        const result = wasm.coppock_js(close, 11, 14, 10, maType);
        assert.strictEqual(result.length, close.length);
        
        
        const warmup = Math.max(11, 14) + (10 - 1);
        
        
        let foundValid = false;
        for (let i = warmup; i < result.length; i++) {
            if (!isNaN(result[i])) {
                foundValid = true;
                break;
            }
        }
        assert(foundValid, `No valid values for MA type ${maType}`);
        results[maType] = result;
    }
    
    
    let allSame = true;
    for (let i = 30; i < 100; i++) {
        if (!isNaN(results['sma'][i]) && !isNaN(results['ema'][i])) {
            if (Math.abs(results['sma'][i] - results['ema'][i]) > 1e-10) {
                allSame = false;
                break;
            }
        }
    }
    assert(!allSame, "Different MA types should produce different results");
});

test('COPPOCK memory leak prevention', () => {
    
    const sizes = [100, 1000, 10000];
    
    for (const size of sizes) {
        const ptr = wasm.coppock_alloc(size);
        assert(ptr !== 0, `Failed to allocate ${size} elements`);
        
        
        const memView = new Float64Array(wasm.__wasm.memory.buffer, ptr, size);
        for (let i = 0; i < Math.min(10, size); i++) {
            memView[i] = i * 1.5;
        }
        
        
        for (let i = 0; i < Math.min(10, size); i++) {
            assert.strictEqual(memView[i], i * 1.5, `Memory corruption at index ${i}`);
        }
        
        
        wasm.coppock_free(ptr, size);
    }
});

test.after(() => {
    console.log('COPPOCK WASM tests completed');
});

File: tests/python/test_coppock.py
"""
Python binding tests for COPPOCK indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

# Add parent directory to path to import the built module
sys.path.insert(0, str(Path(__file__).parent.parent.parent / 'target/wheels'))

try:
    import my_project as ta_indicators
except ImportError:
    pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS


class TestCoppock:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_coppock_accuracy(self, test_data):
        """Test COPPOCK matches expected values from Rust tests"""
        close = test_data['close']
        expected = EXPECTED_OUTPUTS['coppock']
        
        # Use default parameters from expected outputs
        result = ta_indicators.coppock(
            close, 
            expected['default_params']['short'],
            expected['default_params']['long'],
            expected['default_params']['ma']
        )
        
        assert len(result) == len(close)
        
        # Check last 5 values match expected
        for i, expected_val in enumerate(expected['last_5_values']):
            assert_close(result[-(5-i)], expected_val, rtol=1e-8)
    
    def test_coppock_partial_params(self, test_data):
        """Test COPPOCK with default parameters"""
        close = test_data['close']
        
        # Test with default parameters
        result = ta_indicators.coppock(close, 11, 14, 10)
        assert len(result) == len(close)
    
    def test_coppock_default_ma_type(self, test_data):
        """Test COPPOCK with default MA type"""
        close = test_data['close']
        
        # Test without specifying ma_type (should default to "wma")
        result = ta_indicators.coppock(close, 11, 14, 10)
        assert len(result) == len(close)
        
        # Test with explicit ma_type
        result_sma = ta_indicators.coppock(close, 11, 14, 10, "sma")
        assert len(result_sma) == len(close)
    
    def test_coppock_zero_period(self):
        """Test COPPOCK fails with zero period"""
        input_data = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(Exception, match="Invalid period"):
            ta_indicators.coppock(input_data, 0, 14, 10)
    
    def test_coppock_period_exceeds_length(self):
        """Test COPPOCK fails when period exceeds data length"""
        data_small = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(Exception, match="Invalid period"):
            ta_indicators.coppock(data_small, 14, 20, 10)
    
    def test_coppock_very_small_dataset(self):
        """Test COPPOCK fails with insufficient data"""
        single_point = np.array([42.0])
        
        with pytest.raises(Exception, match="Invalid period|Not enough valid data"):
            ta_indicators.coppock(single_point, 11, 14, 10)
    
    def test_coppock_empty_input(self):
        """Test COPPOCK fails with empty input"""
        empty = np.array([])
        
        with pytest.raises(Exception, match="Empty data provided"):
            ta_indicators.coppock(empty, 11, 14, 10)
    
    def test_coppock_all_nan_input(self):
        """Test COPPOCK fails with all NaN values"""
        all_nan = np.full(100, np.nan)
        
        with pytest.raises(Exception, match="All values are NaN"):
            ta_indicators.coppock(all_nan, 11, 14, 10)
    
    def test_coppock_nan_handling(self, test_data):
        """Test COPPOCK handles NaN values correctly"""
        close = test_data['close']
        
        short = 11
        long = 14
        ma = 10
        
        result = ta_indicators.coppock(close, short, long, ma)
        assert len(result) == len(close)
        
        # Warmup period: max(short, long) + (ma - 1)
        warmup = max(short, long) + (ma - 1)  # 14 + 9 = 23
        
        # After warmup period, no NaN values should exist
        if len(result) > warmup:
            non_nan_portion = result[warmup:]
            assert not np.any(np.isnan(non_nan_portion)), f"Found NaN after warmup period {warmup}"
    
    def test_coppock_streaming(self, test_data):
        """Test CoppockStream produces same results as batch"""
        close = test_data['close']
        
        # Batch calculation
        batch_result = ta_indicators.coppock(close, 11, 14, 10)
        
        # Streaming calculation
        stream = ta_indicators.CoppockStream(11, 14, 10, "wma")
        stream_result = []
        
        for price in close:
            value = stream.update(price)
            stream_result.append(value if value is not None else np.nan)
        
        stream_result = np.array(stream_result)
        assert len(stream_result) == len(batch_result)
        
        # Compare results (allowing for NaN equality)
        for i, (b, s) in enumerate(zip(batch_result, stream_result)):
            if np.isnan(b) and np.isnan(s):
                continue
            assert_close(b, s, rtol=1e-7)
    
    def test_coppock_batch_single_param(self, test_data):
        """Test batch with single parameter combination"""
        close = test_data['close']
        
        # Single parameter set: short=11, long=14, ma=10
        batch_result = ta_indicators.coppock_batch(
            close,
            (11, 11, 0),    # short range
            (14, 14, 0),    # long range
            (10, 10, 0)     # ma range
        )
        
        # Should match single calculation
        single_result = ta_indicators.coppock(close, 11, 14, 10)
        
        assert 'values' in batch_result
        batch_values = batch_result['values'][0]  # First row
        
        # Allow tiny FP differences between batch and single.
        # Use absolute tolerance tighter than Rust's 1e-7 reference threshold.
        np.testing.assert_allclose(batch_values, single_result, rtol=0.0, atol=1e-9)
    
    def test_coppock_batch_multiple_params(self, test_data):
        """Test batch with multiple parameter combinations"""
        close = test_data['close'][:100]  # Use smaller dataset for speed
        
        # Multiple parameters
        batch_result = ta_indicators.coppock_batch(
            close,
            (10, 12, 2),    # short: 10, 12
            (14, 16, 2),    # long: 14, 16
            (8, 10, 2)      # ma: 8, 10
        )
        
        assert 'values' in batch_result
        assert 'shorts' in batch_result
        assert 'longs' in batch_result
        assert 'ma_periods' in batch_result
        
        # Should have 2 * 2 * 2 = 8 combinations
        assert batch_result['values'].shape == (8, 100)
        assert len(batch_result['shorts']) == 8
        assert len(batch_result['longs']) == 8
        assert len(batch_result['ma_periods']) == 8
    
    def test_coppock_kernel_selection(self, test_data):
        """Test COPPOCK with different kernel selections"""
        close = test_data['close']
        
        # Test with auto kernel (default)
        result_auto = ta_indicators.coppock(close, 11, 14, 10)
        
        # Test with explicit scalar kernel
        result_scalar = ta_indicators.coppock(close, 11, 14, 10, kernel="scalar")
        
        # Results should be very close
        np.testing.assert_allclose(result_auto, result_scalar, rtol=1e-10)
    
    def test_coppock_invalid_ma_type(self, test_data):
        """Test COPPOCK fails with invalid MA type"""
        close = test_data['close']
        
        # Should raise an error for unknown MA type
        with pytest.raises(Exception, match="Unknown moving average type|Underlying MA error"):
            ta_indicators.coppock(close, 11, 14, 10, "invalid_ma")
    
    def test_coppock_negative_periods(self, test_data):
        """Test COPPOCK fails with negative periods"""
        close = test_data['close']
        
        # Negative short period
        with pytest.raises(OverflowError, match="can't convert negative"):
            ta_indicators.coppock(close, -11, 14, 10)
        
        # Negative long period
        with pytest.raises(OverflowError, match="can't convert negative"):
            ta_indicators.coppock(close, 11, -14, 10)
        
        # Negative MA period
        with pytest.raises(OverflowError, match="can't convert negative"):
            ta_indicators.coppock(close, 11, 14, -10)
    
    def test_coppock_warmup_period(self, test_data):
        """Test COPPOCK warmup period matches expected formula"""
        close = test_data['close']
        
        short = 11
        long = 14
        ma = 10
        
        result = ta_indicators.coppock(close, short, long, ma)
        
        # Warmup period formula: first_valid + max(short, long) + (ma_period - 1)
        # For real data, first_valid is typically 0
        expected_warmup = max(short, long) + (ma - 1)
        
        # Check NaN pattern in warmup period
        for i in range(min(expected_warmup - 1, len(result))):
            assert np.isnan(result[i]), f"Expected NaN at index {i} during warmup"
        
        # Should have valid values after warmup
        if len(result) > expected_warmup:
            assert not np.isnan(result[expected_warmup]), f"Expected valid value at index {expected_warmup}"
    
    def test_coppock_supported_ma_types(self, test_data):
        """Test COPPOCK with supported MA types"""
        close = test_data['close'][:100]  # Use smaller dataset
        
        # Only test MA types that are actually supported
        supported_ma_types = ['sma', 'ema', 'wma']
        
        results = {}
        for ma_type in supported_ma_types:
            result = ta_indicators.coppock(close, 11, 14, 10, ma_type)
            assert len(result) == len(close)
            # Each MA type should produce different results
            assert not np.all(np.isnan(result))
            results[ma_type] = result
        
        # Verify different MA types produce different results
        assert not np.allclose(results['sma'], results['ema'], equal_nan=True)
        assert not np.allclose(results['wma'], results['ema'], equal_nan=True)


if __name__ == '__main__':
    pytest.main([__file__, '-v'])
